libparsec/crates/client/src/certif/realms_needs.rs

```diff
@@ -0,0 +1,115 @@
+
+#[derive(Debug, thiserror::Error)]
+pub enum CertifGetRealmNeedsError {
+    #[error("Component has stopped")]
+    Stopped,
+    #[error(transparent)]
+    Internal(#[from] anyhow::Error),
+}
+
+impl From<CertifStoreError> for CertifGetRealmNeedsError {
+    fn from(value: CertifStoreError) -> Self {
+        match value {
+            CertifStoreError::Stopped => Self::Stopped,
+            CertifStoreError::Internal(err) => err.into(),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum RealmNeeds {
+    Nothing,
+    KeyRotationOnly {
+        current_key_index: Option<IndexInt>,
+    },
+    /// A user that is part of the workspace has been revoked, this creates a cascade
+    /// of needs: first the workspace needs to be unshared, which is turn means a key
+    /// rotation is needed.
+    UnshareThenKeyRotation {
+        current_key_index: Option<IndexInt>,
+        revoked_users: Vec<UserID>,
+    },
+}
+
+/// Note realm existence is not checked (and `RealmNeeds::Nothing` is returned if that's not the case)
+pub async fn get_realm_needs(
+    ops: &CertificateOps,
+    realm_id: VlobID,
+) -> Result<RealmNeeds, CertifGetRealmNeedsError> {
+    ops.store
+        .for_read(|store| async move {
+            // TODO: This implementation is not efficient, as it fetches all the roles
+            //       and users certificates. However this should be good enough for now,
+            //       and optimization requires a complex API changes in the storage.
+
+            let mut needs = RealmNeeds::Nothing;
+
+            let last_key_rotation = store
+                .get_realm_last_key_rotation_certificate(UpTo::Current, realm_id)
+                .await?;
+            let current_key_index = last_key_rotation.as_ref().map(|x| x.key_index);
+
+            let roles = store.get_realm_roles(UpTo::Current, realm_id).await?;
+            let mut current_users_with_role = HashSet::new();
+            // This loop does two things at once:
+            // - Collect all users that are currently part of the workspace (will be used in next step)
+            // - Update `needs` if a unshare has occured since the last key rotation
+            for role in roles {
+                match role.role {
+                    Some(_) => {
+                        current_users_with_role.insert(role.user_id);
+                    }
+                    None => {
+                        current_users_with_role.remove(&role.user_id);
+
+                        match &last_key_rotation {
+                            Some(last_key_rotation)
+                                if last_key_rotation.timestamp < role.timestamp =>
+                            {
+                                needs = RealmNeeds::KeyRotationOnly { current_key_index };
+                            }
+                            None => {
+                                needs = RealmNeeds::KeyRotationOnly { current_key_index };
+                            }
```

Comment:
nit: You could maybe merge these two branches

Reply:
I can't because there is the if clause that is only for the first branch
```rust
enum X {
  A(u32),
  B(u32),
  C
}

match X::A(1) {
  X::A(x) | X::B(x) if x == 1 => (),  // This is allowed
  X::C | X::B(x) if x == 1 => (), // C doesn't bind x, this cannot compile
   _ => (),
}
```

See [this playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8be3edeae62afef17af63388d6fa8b5e)